config/main.py

```diff
@@ -3153,111 +3214,110 @@ def del_vrf_vni_map(ctx, vrfname):
-        ctx.fail("argument is not in pattern prefix [vrf <vrf_name>] <A.B.C.D/M> nexthop <[vrf <vrf_name>] <A.B.C.D>>|<dev <dev_name>>!")
-    if "prefix" not in command_str:
-        ctx.fail("argument is incomplete, prefix not found!")
-    if "nexthop" not in command_str:
-        ctx.fail("argument is incomplete, nexthop not found!")
-    for i in range(0, len(command_str)):
-        if "nexthop" == command_str[i]:
-            prefix_str = command_str[:i]
-            nexthop_str = command_str[i:]
-    vrf_name = ""
-    cmd = 'sudo vtysh -c "configure terminal" -c "ip route'
-    if prefix_str:
-        if len(prefix_str) == 2:
-            prefix_mask = prefix_str[1]
-            cmd += ' {}'.format(prefix_mask)
-        elif len(prefix_str) == 4:
-            vrf_name = prefix_str[2]
-            prefix_mask = prefix_str[3]
-            cmd += ' {}'.format(prefix_mask)
-        else:
-            ctx.fail("prefix is not in pattern!")
-    if nexthop_str:
-        if len(nexthop_str) == 2:
-            ip = nexthop_str[1]
-            if vrf_name == "":
-                cmd += ' {}'.format(ip)
-            else:
-                cmd += ' {} vrf {}'.format(ip, vrf_name)
-        elif len(nexthop_str) == 3:
-            dev_name = nexthop_str[2]
-            if vrf_name == "":
-                cmd += ' {}'.format(dev_name)
-            else:
-                cmd += ' {} vrf {}'.format(dev_name, vrf_name)
-        elif len(nexthop_str) == 4:
-            vrf_name_dst = nexthop_str[2]
-            ip = nexthop_str[3]
-            if vrf_name == "":
-                cmd += ' {} nexthop-vrf {}'.format(ip, vrf_name_dst)
+    config_db = ctx.obj['config_db']
+    key, route = cli_sroute_to_config(ctx, command_str)
+
+    # If defined intf name, check if it exists
+    if 'ifname' in route:
+        if (not route['ifname'] in config_db.get_keys('VLAN') and
+            not route['ifname'] in config_db.get_keys('INTERFACE') and
+            not route['ifname'] in config_db.get_keys('PORTCHANNEL')):
+            ctx.fail('interface {} doesn`t exist'.format(route['ifname']))
+
+    # Check if exist entry with key
+    keys = config_db.get_keys('STATIC_ROUTE')
+    if key in keys:
+        # If exist update current entry
+        current_entry = config_db.get_entry('STATIC_ROUTE', key)
+
+        for entry in ['nexthop', 'nexthop-vrf', 'ifname']:
+            if not entry in current_entry:
+                    current_entry[entry] = ''
+            if entry in route:
+                current_entry[entry] += ',' + route[entry]
             else:
-                cmd += ' {} vrf {} nexthop-vrf {}'.format(ip, vrf_name, vrf_name_dst)
-        else:
-            ctx.fail("nexthop is not in pattern!")
-    cmd += '"'
-    clicommon.run_command(cmd)
+                current_entry[entry] += ','
+
+        config_db.set_entry("STATIC_ROUTE", key, current_entry)
+    else:
+        config_db.set_entry("STATIC_ROUTE", key, route)
```

Comment:
I think there is an assumption in this design that the static routes are always configured by this CLI command. I think this assumption may not always be true. There are a few scenarios that might be problematic. Here are a few examples:
1. The original `"STATIC_ROUTE":{"1.1.1.1/24": {"nexthop": "2.2.2.2,3.3.3.3"}}` (I know this should never happen with this CLI command only, but it might be valid using some other ways since it is valid according to the schema). If we add an additional nexthop `4.4.4.4`, then with the current logic the entry would become `"STATIC_ROUTE":{"1.1.1.1/24": {"nexthop": "2.2.2.2,3.3.3.3,4.4.4.4", "ifname": ",", "nexthop-vrf": ","}}`. This would be invalid since `nexthop` has three values while others have only two.
2. The original table has some other fields (e.g., distance, blackhole). This command would never modify those fields and the numbers would not match after executing this command.

Similar situations need to be considered in del as well.

Reply:
IMO, scenario 1 could be low prioritized as it wouldn't be a common use-case. @d-dashkov , can you please confirm 2?

Reply:
After running this command, the numbers might not match, and to prevent this, I added the creation of these fields and set their default values according to https://github.com/project-arlo/SONiC/blob/1e8551e0ffe6a01ac5a717491814943d37e29ad9/doc/static -route / SONiC_static_route_route .md # mapping-from-table-entry-to-frr-command-arguments, so now user can manually change values in config_db